modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/FailureStoreMetricsWithIncrementalBulkIT.java

```diff
@@ -0,0 +1,283 @@
+        String backingIndex = getDataStreamResponse.getDataStreams().get(0).getDataStream().getIndices().get(0).getName();
+        assertThat(backingIndex, backingIndexEqualTo(dataStream, 1));
+
+        String node = findShard(resolveIndex(backingIndex), 0);
+        IndexingPressure primaryPressure = internalCluster().getInstance(IndexingPressure.class, node);
+        long memoryLimit = primaryPressure.stats().getMemoryLimit();
+        long primaryRejections = primaryPressure.stats().getPrimaryRejections();
+        try (Releasable releasable = primaryPressure.markPrimaryOperationStarted(10, memoryLimit, false)) {
+            while (primaryPressure.stats().getPrimaryRejections() == primaryRejections) {
+                while (nextRequested.get()) {
+                    nextRequested.set(false);
+                    refCounted.incRef();
+                    List<DocWriteRequest<?>> requests = new ArrayList<>();
+                    for (int i = 0; i < 20; ++i) {
+                        requests.add(indexRequest(dataStream));
+                    }
+                    handler.addItems(requests, refCounted::decRef, () -> nextRequested.set(true));
+                }
+                assertBusy(() -> assertTrue(nextRequested.get()));
+            }
+        }
+
+        while (nextRequested.get()) {
+            nextRequested.set(false);
+            refCounted.incRef();
+            handler.addItems(List.of(indexRequest(dataStream)), refCounted::decRef, () -> nextRequested.set(true));
+        }
+
+        assertBusy(() -> assertTrue(nextRequested.get()));
+
+        PlainActionFuture<BulkResponse> future = new PlainActionFuture<>();
+        handler.lastItems(List.of(indexRequest(dataStream)), refCounted::decRef, future);
+
+        BulkResponse bulkResponse = safeGet(future);
+
+        /* The datastream should have a failure store associated with it now */
+        getDataStreamRequest = new GetDataStreamAction.Request(TEST_REQUEST_TIMEOUT, new String[] { "*" });
+        getDataStreamResponse = client().execute(GetDataStreamAction.INSTANCE, getDataStreamRequest).actionGet();
+        Index failureStoreIndex = getDataStreamResponse.getDataStreams().get(0).getDataStream().getFailureStoreWriteIndex();
+        String failureStoreIndexName = failureStoreIndex.getName();
+        String failure_node = findShard(resolveIndex(failureStoreIndexName), 0);
+        boolean shardsOnDifferentNodes = node.equals(failure_node) == false;
+
+        for (int i = 0; i < hits.get(); ++i) {
+            assertFalse(bulkResponse.getItems()[i].isFailed());
+            assertTrue(bulkResponse.getItems()[i].getFailureStoreStatus().getLabel().equalsIgnoreCase("NOT_APPLICABLE_OR_UNKNOWN"));
+        }
+
+        int docs_redirected_to_fs = 0;
+        int docs_in_fs = 0;
+        for (int i = (int) hits.get(); i < bulkResponse.getItems().length; ++i) {
+            BulkItemResponse item = bulkResponse.getItems()[i];
+            if (item.isFailed()) {
```

Comment:
Just so I understand this correctly:

If an indexing operation fails and there is a failure store configured will then try to store that failure in the failure store. If that operation succeeds, then the indexing operations is indicated to the user as "successful"?

As long as that is the expected behavior this mitigation looks good to me.

Reply:
Yes your understanding is correct @Tim-Brooks. If we successfully index into the failure store, the indexing operation is considered successful, even though the operation failed to index into the original index.

Reply:
Hi @ankikuma & @Tim-Brooks ,

> As long as that is the expected behavior this mitigation looks good to

I do not think this is expected behaviour. 

I am afraid I mislead @ankikuma during our last chat.

> If we successfully index into the failure store, the indexing operation is considered successful, even though the operation failed to index into the original index.

This is correct, and it explains the behaviour that the code exhibits. But if I remember correctly the conversations we looked into, the purpose of the failure store is to store failures as a result from a user misconfiguration and not technical limitations/failures. 

Considering this, I would expect that this type of failure should not be redirected to the failure store and it should result in a failed response.

As a way forward, I see two options depending on the scope of this work:
- If the purpose of this test is to check that incremental indexing and failure store work as expected, I would say that we need to fix the bug that this test has unearthed.
- If the purpose is to only add a test to cover the failure store and incremental bulk indexing working together, we should write the test to work as expected and then open a bug and mute this test.

Preferably, I would prefer the first but I do not fully know the scope of this work, so I would like to offer an alternative as well.

Reply:
@gmarouli The scope of this test was to just test the interaction of incremental bulk failures caused by the short circuit feature with failure store, based on the comment from @nielsbauman [here](https://github.com/elastic/elasticsearch/pull/113044#pullrequestreview-2312162329). And in this original comment , it looks like we expected the short circuited requests to got to the failure store.

However, as this [discussion](https://elastic.slack.com/archives/CE8P96STG/p1726684093864779?thread_ts=1726598358.836249&cid=CE8P96STG) suggests, James Baiera indicated that if failures occurred due to resource constraints, they should not go to the failure store.

Now a short circuit failure is triggered due to a previous failure. It just so happens that in this test we simulate that failure based on indexing pressure. I am not sure how one would distinguish between failures caused by resource constraints vs. other types of failures (for the benefit of the failure store).